utils/controller_debug.py

```diff
@@ -0,0 +1,1058 @@
+            print("Please give the variable name and value")
+            return
+        SetVar(cl[0], cl[1])
+
+    def complete_set(self, text, line, begidx, endidx):
+        var = text
+
+        out = []
+        for i in varDict.keys():
+            if i.startswith(var):
+                out.append(i)
+
+        return out
+
+    def help_set(self):
+        global varDict
+        print("Set the value of a debug variable\n")
+        print("Variables currently defined:")
+        for k in varDict.keys():
+            print("   %-10s - %s" % (k, varDict[k].help))
+
+    def do_trace(self, line):
+        """
+       This command is used to access the trace buffer on the controller.
+       A sub-command must be passed as an option:
+
+       trace flush
+         Stops the trace if one was on-going and flushes the trace buffer
+
+       trace start
+         Starts the trace collecting data.
+
+       trace graph
+         This will download the data and display it graphically
+
+       trace download <filename>
+         This will download the data and save it to a file with the given
+         name.  If no file name is given, then trace.dat will be used
+
+       The trace feature allows variables to be sampled in real time and saved
+       to a large internal memory buffer in the drive.  This command can then
+       be used to download the captured data and save or display it.
+
+       The trace first must be set up by setting specific variables.
+
+       trace_var1
+       trace_var2
+       trace_var3
+       trace_var4
+         Set these variables to the name of the parameter you want to capture.
+         Any parameter can be given by name.  Up to 4 parameters can be saved
+         to the trace buffer at any time.  Think of it as a 4 channel oscilloscope.
+
+       trace_period
+         Set this to the period at which you want to save the data.  The units
+         are loop periods.  A period <= 0 is treated as 1
+
+       trace_ctrl
+         This variable is used to start the trace.  Just set it to 1 to start.
+         It will be cleared when the trace buffer is full.
+
+       trace_samples
+         This variable gives the number of samples stored in the buffer currently.
+       """
+        cl = line.split()
+        if len(cl) < 1:
+            print("Error, please specify the trace command to run")
+            return
+
+        if cl[0] == "flush":
+            SendCmd(OP_TRACE, [SUBCMD_TRACE_FLUSH])
+
+        elif cl[0] == "start":
+            SendCmd(OP_TRACE, [SUBCMD_TRACE_FLUSH])
+            SetVar("trace_ctrl", 1)
+
+        elif cl[0] == "download":
+            tv = TraceActiveVars()
+            if len(tv) < 1:
+                print("No active trace variables")
+
+            if len(cl) > 1:
+                fname = cl[1]
+            else:
+                fname = "trace.dat"
+            fp = open(fname, "w")
+            dat = TraceDownload()
+            for i in range(len(dat[0])):
```

Comment:
Would it be a good idea to print a header into the file, containing the names of variables being traced?

Reply:
Yes, in fact I was planning to do that and never got around to it.  Will do.